src/core/Track.cpp

```diff
@@ -1125,25 +1165,170 @@ void TrackContentObjectView::contextMenuEvent( QContextMenuEvent * cme )
-						this, SLOT( remove() ) );
+					tr( individualTCO
+						? "Delete (middle mousebutton)"
+						: "Delete selection (middle mousebutton)" ),
+						[this](){ contextMenuAction( Remove ); } );
 		contextMenu.addSeparator();
 		contextMenu.addAction( embed::getIconPixmap( "edit_cut" ),
-					tr( "Cut" ), this, SLOT( cut() ) );
+					tr( individualTCO
+						? "Cut"
+						: "Cut selection" ),
+						[this](){ contextMenuAction( Cut ); } );
 	}
 	contextMenu.addAction( embed::getIconPixmap( "edit_copy" ),
-					tr( "Copy" ), m_tco, SLOT( copy() ) );
+					tr( individualTCO
+						? "Copy"
+						: "Copy selection" ),
+						[this](){ contextMenuAction( Copy ); } );
 	contextMenu.addAction( embed::getIconPixmap( "edit_paste" ),
-					tr( "Paste" ), m_tco, SLOT( paste() ) );
+					tr( "Paste" ), [this](){ contextMenuAction( Paste ); } );
 	contextMenu.addSeparator();
 	contextMenu.addAction( embed::getIconPixmap( "muted" ),
-				tr( "Mute/unmute (<%1> + middle click)" ).arg(UI_CTRL_KEY),
-						m_tco, SLOT( toggleMute() ) );
+				tr( individualTCO
+					? "Mute/unmute (<%1> + middle click)"
+					: "Mute/unmute selection (<%1> + middle click)" ).arg(UI_CTRL_KEY),
+					[this](){ contextMenuAction( Mute ); } );
 	constructContextMenu( &contextMenu );
 
 	contextMenu.exec( QCursor::pos() );
 }
 
+// This method processes the actions from the context menu of the TCO View.
+void TrackContentObjectView::contextMenuAction( ContextMenuAction action )
+{
+	// Get a list of selected selectableObjects
+	QVector<selectableObject *> sos = gui->songEditor()->m_editor->selectedObjects();
+
+	// Convert to a list of selected TCOVs
+	QVector<TrackContentObjectView *> selection;
+	selection.reserve( sos.size() );
+	for( auto so: sos )
+	{
+		TrackContentObjectView *tcov = dynamic_cast<TrackContentObjectView *> ( so );
+		if( tcov != nullptr )
+		{
+			selection.append( tcov );
+		}
+	}
+
+	// If we clicked part of the selection, affect all selected clips. Otherwise affect the clip we clicked
+	QVector<TrackContentObjectView *> active = selection.contains( this ) ? selection : QVector<TrackContentObjectView *>( 1, this );
+	// active will be later used for the removeActive, copyActive, cutActive or toggleMuteActive methods
+
+	switch( action )
+	{
+		case Remove:
+			removeActive( active );
+			break;
+		case Cut:
+			cutActive( active );
+			break;
+		case Copy:
+			copyActive( active );
+			break;
+		case Paste:
+			// NOTE: Because we give preference to the QApplication clipboard over the LMMS Clipboard class, we need to
+			// clear the QApplication Clipboard during the LMMS Clipboard copy operations (Clipboard::copy does that)
+
+			// If we have TCO data on the clipboard paste it. If not, do our regular TCO paste.
+			if( QApplication::clipboard()->mimeData( QClipboard::Clipboard )->hasFormat( StringPairDrag::mimeType() ) )
+			{
+				pasteSelection();
+			}
+			else
+			{
+				getTrackContentObject()->paste();
+			}
```

Comment:
What are your opinions on this logic staying here? Should I move it to the `pasteSelection` method and change its name to something like `pasteContent`? Or maybe keep it that way and just move everything to the `paste` method once we (possibly) move away from using `Clipboard::copy`?